networkx/algorithms/approximation/fvs.py

```diff
@@ -0,0 +1,222 @@
+    """
+
+    def move_and_cleanup(G, F, weights, stack=None):
+        """Move all nodes with weight zero from G to F.
+
+        Each node of weight zero moved from `G` to `F` is recorded in
+        `stack`, if given. After removing the nodes of weight zero from
+        `G` and adding them to `F`, this function modifies `G` to be its
+        two-core. This also clears the nodes of weight zero from the
+        `weights` dictionary as well.
+
+        """
+        # Move all weight zero nodes from G to F, pushing each onto the
+        # stack if necessary.
+        weight_zero_nodes = {v for v in G if weights[v] == 0}
+        G.remove_nodes_from(weight_zero_nodes)
+        F |= weight_zero_nodes
+        if stack is not None:
+            stack.extend(weight_zero_nodes)
+        # Compute the two-core and remove each node not in the two-core
+        # from the dictionary of weights.
+        before = set(G)
+        G = nx.k_core(G, k=2)
+        for v in before - set(G):
+            del weights[v]
+
+    # Store the original graph, since G will be modified.
+    G_original, G = G, G.copy(with_data=False)
+    # Create a separate dictionary of weights, since the weights of each
+    # node will be updated in the main loop.
+    weights = Counter({u: G.node[u].get(weight, 1) for u in G})
+    # Initially, F is the empty set. As the algorithm progresses, nodes
+    # will be removed from the graph and added to F. At the end of the
+    # algorithm, F will contain a feedback vertex set.
+    F = set()
+    # The stack stores the order in which nodes are removed from the graph.
+    stack = []
+    # First, remove all nodes of weight zero from the graph and add them
+    # to F. Clean up the graph by removing any dangling tails.
+    move_and_cleanup(G, F, weights)
+    while G:
+        # Try to find a semidisjoint cycle in the graph.
+        try:
+            C = next(semidisjoint_cycles(G))
+        # If there are no semidisjoint cycles, update the degree-scaled
+        # weights of all nodes.
+        except StopIteration:
+            gamma = min(weights[v] / (d - 1) for v, d in G.degree())
+            weights -= {v: gamma * (d - 1) for v, d in G.degree()}
+        # Otherwise, only update the weights of the nodes in the cycle.
+        else:
+            gamma = min(weights[v] for v in C)
+            weights -= {v: gamma for v in C}
+        # Remove all nodes of weight zero from the graph and add them to
+        # F. Clean up the graph by removing any dangling tails.
+        move_and_cleanup(G, F, weights, stack=stack)
+    # Remove extraneous nodes from the feedback vertex set. A node v is
+    # extraneous to the feedback vertex set F if F - {v} remains a
+    # feedback vertex set in the original graph. This not only ensures a
+    # minimal feedback vertex set but is also required to achieve the
+    # approximation guarantee.
+    while stack:
+        v = stack.pop()
+        if nx.is_forest(G_original.subgraph(F - {v})):
```

Comment:
Here we're checking whether the FVS minus vertex v is a forest - but shouldn't we actually be checking whether the original graph minus (FVS - v) is a forest?